### ISSUE TITLE
[Bug] [zos_job_submit] 'utf-8' codec can't decode byte 0xbb in position 0: invalid start byte"

### ISSUE THREAD

**Issue by rishijoshi07** [ansible-collections/ibm_zos_core#972]
### Is there an existing issue for this?

- [X] There are no existing issues.

### Are the dependencies a supported version?

- [X] The dependencies are supported.

### IBM Z Open Automation Utilities

v1.2.3

### IBM Enterprise Python

v3.10.x

### IBM z/OS Ansible core Version

v1.7.0

### ansible-version

v2.14.x

### z/OS version

v2.5

### Ansible module

zos_job_submit

### Bug description

When I try to run the playbook I got "UTF-8' codec can't decode byte 0xbb in position 0: invalid start byte",
Can you please provide the solution?

### Playbook verbosity output.

{
  "dest": "/tmp/ansible.ofznxdp1",
  "src": "/tmp/ansible-tmp-1694759015.884247-51-1064403726060/source",
  "md5sum": "d62a61a4557e7fed5b4abc7f3afcd56e",
  "checksum": "a28f35a17962351fc326fc4331fa3eac82d4d4c2",
  "changed": false,
  "uid": 266,
  "gid": 0,
  "owner": "XXXXXXXX",
  "group": "XXXXXXXX",
  "mode": "0600",
  "state": "file",
  "size": 7077,
  "invocation": {
    "module_args": {
      "src": "./files/SYSJOB1A.jcl",
      "location": "LOCAL",
      "wait": true,
      "wait_time_s": 60,
      "encoding": {
        "from": "ISO8859-1",
        "to": "IBM-1047"
      },
      "temp_file": "/tmp/ansible.ofznxdp1",
      "return_output": true,
      "volume": null,
      "max_rc": null,
      "from_encoding": "ISO8859-1",
      "to_encoding": "IBM-1047"
    }
  },
  "msg": "The JCL submitted with job id JOB06434 but there was an error, please review the error for further details: 'utf-8' codec can't decode byte 0xbb in position 0: invalid start byte",
  "deprecations": [
    {
      "msg": "Param 'wait' is deprecated. See the module docs for more information",
      "date": "2022-11-30",
      "collection_name": "ibm.ibm_zos_core"
    }
  ],
  "_ansible_no_log": null
}

**Comment by rishinz** [comment]
Hi @ketankelkar @ddimatos - could you please provide your suggestion on this. Any help is appreciated.

Thanks!

**Comment by rishijoshi07** [comment]
This is my playbook


---
- hosts: LPAR
  collections:
    - ibm.ibm_zos_core
  gather_facts: no
  vars:
  environment: "{{ environment_vars }}"


  tasks:
    - name: CSQ2PARM Job template Render
      local_action:
        module: ansible.builtin.template
        src: './files/CSQ2PARM.j2'
        dest: './files/CSQ2PARM.jcl'
      register: result

    - name: CSQ2PARM Job Template Response
      debug:
        msg: "{{ result }}"

    - name: Convert local JCL file to IBM-1047 and Submit the job
      zos_job_submit:
        src: './files/CSQ2PARM.jcl'
        location: LOCAL
        wait: true
        wait_time_s: 60
        encoding:
          from: ISO8859-1
          to: IBM-037
      register: result

    - name: CSQ2PARM Job Submission Response
      debug:
        msg: "{{ result }}"

    - name: Job Response
      debug:
        msg: "{{ result.jobs[0].ret_code }}" 
        
        
I have tried many scenarios to encode UTF-8 to IBM-037 and IBM-1047 but nothing works.
Can you please give me a suggestion?

**Comment by richp405** [comment]
See also JIRA 9687

**Comment by richp405** [comment]
see #677 for recreation

**Comment by ddimatos** [comment]
@rishijoshi07 - looking over the issue, it looks like it’s the same issue reported in #677 which really has nothing to do with the encoding used during job submission but instead the result of the job submissions output resulting in non-printable characters in the job output. When a job result is read by our C and python api's they expect characters to correspond to the UTF-8 charset and when they don't, python has no way of interacting with that character such as printing it to stdout. This also explained in i issue #677. 

Each work around is unique in its own, while we do have a prototype that will sanitize output before it reaches our modules python (prior to ibm_zos_core 1.5.0 - we used REXX and this did not occur, only with the more rigid python and C is this occurring now), this prototype will beta release in Q1 2024.

Some work arounds I have heard depend on the operations in the JCL, for example.
- Specify that the ASA assembler option be enabled that instructs the assembler to use ANSI control characters instead of machine code control characters in the assembler listing if your job is interacting with [additional RECFM's](https://www.ibm.com/docs/en/zos/2.5.0?topic=io-record-formats) ending in `M` that specifies the file contains machine control characters.
- Another option is to add `ignore_errors:true` to the `zos_job_submit` task so the task error you see will not fail the playbook
- Building on the second option:
   - 'Another option is to add `ignore_errors:true` to the `zos_job_submit` task so the task error you see will not fail the playbook' and capture the result of the failing `zos_job_submit` task by registering a variable
   - then in a `set_fact` task, parse the `zos_job_submit` with a regular expression to obtain the job id for the response that has a `0xbb in position 0` in it, now you have the job id.
   - then use `zos_job_output` to read the particular **non-offending DD**, what i mean by this is your error is `0xbb in position 0` this the non-printable char is at position 0 of the jobs output, if you look at your job in hex, i bet the hex value will not correspond to a printable UTF-8 char, there could be more than one non-printable, but most often it’s the first one used by the system to identify the a start byte. Anyhow, with this step, you can go use `zos_job_output` to read just dd you want, one without a non-printable char to see if your job was successful such as `JESMSGLG`, or another DD.

I hope this will help you with some possible work arounds.

**Comment by ddimatos** [comment]
See the following issues below, they are all related to non-printable UTF-8 chars, essentially chararcetors that don't don't correspond to a UTF-8 value. This is being addressed as noted in ZOAU 1.3 and IBM z/OS Core collection will adopt ZOAU 1.3 in Q1 2024. There is a recreate involving a job submission you can see [here](https://github.com/ansible-collections/ibm_zos_core/issues/677#issuecomment-1666774931). This work is also being tracked in JIRA 9687.
- #677
- #776
- #972

**Comment by jacobemery** [comment]
For those looking for a temporary workaround, I did some troubleshooting and [this playbook](https://github.com/jacobemery/aap4zos/blob/main/zos_concepts/jobs/submit_query_retrieve/job_submit_any.yml) is working for me:

- The playbook submits JCL from an Ansible Automation Platform survey "job_contents_survey"
- Then copies the contents to z/OS and prints it for visual confirmation that it was transferred properly
- Submits the job, ignoring errors initially so that output can be captured in the event of a failure
- Gathers the job ID using regex as @ddimatos suggested above
- Attempts to gather the output using the job ID
- If gathering the output fails, it drops to the rescue tasks and prints the error message, and then fails the playbook for accurate results
- If gathering the output succeeds, it prints the full output of the job and the playbook succeeds

Of course you could also submit the JCL via a file instead of with an AAP survey. If you do want to use AAP and a survey, here's what I did:

Here's the AAP job template definition:
```
    - name: Create 'z/OS Job Submit Any' job template.
      tags: job_template, template, zos_job_submit_any, job_submit_any
      ansible.controller.job_template:
        controller_oauthtoken: "{{ controller_token }}"
        name: "z/OS Job Submit Any"
        description: "A playbook for submitting any JCL directly to the z/OS host from AAP survey contents."
        job_type: "run"
        organization: "Default"
        inventory: "AAP z/OS"
        project: "aap4zos"
        playbook: "zos_concepts/jobs/submit_query_retrieve/job_submit_any.yml"
        credentials: "z/OS Host SSH Key"
        execution_environment: "aap4zos EE"
        state: "present"
        validate_certs: false
        survey_enabled: true
        survey_spec: "{{ lookup('file', 'submit_job_any_survey_spec.json') }}"
```

The last parameter points to this AAP survey definition:
```
{
    "name": "z/OS Job Submit Any",
    "description": "A playbook for submitting any JCL directly to the z/OS host from AAP survey contents.",
    "spec": [
        {
        "type": "text",
        "question_name": "Job Name",
        "question_description": "A name for the job to submit.",
        "variable": "job_name",
        "required": true
        },
        {
        "type": "textarea",
        "question_name": "Job Contents",
        "question_description": "JCL contents of the job to submit.",
        "variable": "job_contents",
        "required": true
        }
    ]
}
```

Hope this helps anyone in a similar situation!
